Tear down pending HTTP connection when the originating request completes

antonfirsov opened the pull request:
Resolves #66297

/cc @stephentoub

antonfirsov committed:
cancel pending connections when request is cancelled

antonfirsov committed:
delete old methods

antonfirsov committed:
move CancelPendingRequest_DropsStalledConnectionAttempt_WithGlobalTimeout to outer loop

antonfirsov committed:
comments

antonfirsov committed:
language

antonfirsov committed:
language

antonfirsov commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:
```diff
@@ -2419,5 +2439,31 @@ public QueueItem PeekNextRequestForConnectionAttempt()
             public int RequestsWithoutAConnectionAttempt => _size - _attemptedConnectionsOffset;
         }
+
+        private sealed class HttpConnectionWaiter<T> : TaskCompletionSourceWithCancellation<T>
```
There are no other usages of `TaskCompletionSourceWithCancellation<T>`. Should we merge these types?

stephentoub replied:
@greenEkatherine is adding another use in a PR.

antonfirsov committed:
minor simplification

antonfirsov committed:
attempt to fix CA1852 on wasm targets

antonfirsov committed:
better logging in CheckForHttp*ConnectionInjection

antonfirsov committed:
cleanup

stephentoub commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:
```diff
@@ -669,27 +664,32 @@ private async Task AddHttp2ConnectionAsync(RequestQueue<Http2Connection?>.QueueI
                 catch (Exception e)
                 {
-                    HandleHttp2ConnectionFailure(queueItem.Waiter, e);
+                    waiter.ConnectionCancellationTokenSource = null;
```
Same comments/questions as earlier.

stephentoub commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:
```diff
@@ -733,22 +746,22 @@ private void CheckForHttp2ConnectionInjection()
             // Look for a usable connection.
-            TaskCompletionSourceWithCancellation<Http2Connection?> waiter;
+            waiter = null;
+            connection = null;
```
Ditto

stephentoub commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:
```diff
@@ -1095,6 +1104,31 @@ public async ValueTask<HttpResponseMessage> SendWithVersionDetectionAndRetryAsyn
+                }
+                catch (ObjectDisposedException)
+                {
+                    // There is a low-probability race for the CTS being disposed by the time of the cancellation call, ignore this ODE.
```
It's not thread-safe to CTS.Cancel{After} and CTS.Dispose concurrently.  Further, if you Dispose it after having called CancelAfter, you won't end up canceling it as you'll be disposing of the timer that would have eventually done so.

antonfirsov committed:
Merge branch 'main' into cancel-stalled-connections

antonfirsov committed:
nits

antonfirsov committed:
synchronize access to ConnectionCTS

antonfirsov committed:
TryGetPooledHttp2Connection out params

antonfirsov committed:
add comment on CancelAfter

antonfirsov commented:
/azp run runtime-libraries-coreclr outerloop

antonfirsov commented:
/azp run runtime-libraries stress-http

antonfirsov committed:
skip CancelPendingRequest_DropsStalledConnectionAttempt on HTTP/3 for now

antonfirsov commented:
@stephentoub @dotnet/ncl this is ready for another round of reviews.

stephentoub commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:
```diff
@@ -647,64 +650,82 @@ private async Task AddHttp2ConnectionAsync(RequestQueue<Http2Connection?>.QueueI
+            }
+            finally
+            {
+                waiter.ConnectionCancellationTokenSource = null;
```
Ditto

stephentoub commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:
```diff
@@ -647,64 +650,82 @@ private async Task AddHttp2ConnectionAsync(RequestQueue<Http2Connection?>.QueueI
+            }
+            catch (OperationCanceledException oce) when (oce.CancellationToken == cts.Token)
+            {
+                connectionException = CreateConnectTimeoutException(oce);
```
Ditto

stephentoub commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:
```diff
@@ -1095,6 +1127,30 @@ public async ValueTask<HttpResponseMessage> SendWithVersionDetectionAndRetryAsyn
+
+            lock (waiter)
+            {
+                if (waiter.ConnectionCancellationTokenSource == null) return;
+
+                if (NetEventSource.Log.IsEnabled())
+                    pool.Trace($"Cancelling a pending connection attempt with timeout of {GlobalHttpSettings.SocketsHttpHandler.PendingConnectionTimeoutOnRequestCompletion} ms");
+
+                if (GlobalHttpSettings.SocketsHttpHandler.PendingConnectionTimeoutOnRequestCompletion > 0)
+                    // This cancellation will not fire if the connection succeeds within the delay and the CTS becomes disposed.
+                    waiter.ConnectionCancellationTokenSource.CancelAfter(GlobalHttpSettings.SocketsHttpHandler.PendingConnectionTimeoutOnRequestCompletion);
+                else
+                    waiter.ConnectionCancellationTokenSource.Cancel(); // Cancel immediately by default
```
```suggestion
                if (waiter.ConnectionCancellationTokenSource == null)
                {
                    return;
                }

                if (NetEventSource.Log.IsEnabled())
                {
                    pool.Trace($"Cancelling a pending connection attempt with timeout of {GlobalHttpSettings.SocketsHttpHandler.PendingConnectionTimeoutOnRequestCompletion} ms");
                }

                if (GlobalHttpSettings.SocketsHttpHandler.PendingConnectionTimeoutOnRequestCompletion > 0)
                {
                    // Cancel after the specified timeout. This cancellation will not fire if the connection
                    // succeeds within the delay and the CTS becomes disposed.
                    waiter.ConnectionCancellationTokenSource.CancelAfter(GlobalHttpSettings.SocketsHttpHandler.PendingConnectionTimeoutOnRequestCompletion);
                }
                else
                {
                    // Cancel immediately by default
                    waiter.ConnectionCancellationTokenSource.Cancel();
                }
```

stephentoub commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:
```diff
@@ -1095,6 +1127,30 @@ public async ValueTask<HttpResponseMessage> SendWithVersionDetectionAndRetryAsyn
                 }
+                finally
+                {
+                    CancelWithGlobalTimeout(this, http11ConnectionWaiter);
+                    CancelWithGlobalTimeout(this, http2ConnectionWaiter);
```
In what situation would both of these be in need of cancellation?  If you nulled them out upon successful completion of waiting for the request, could this instead be:
```C#
CancelWithGlobalTimeout(this, http11ConnectionWaiter ?? http2ConnectionWaiter);
```
?

Also, as a nit, can we name this CancelIfNecessary or something like that?  It's not always going to use a timeout, and there may not even be anything to cancel.

antonfirsov replied:
> In what situation would both of these be in need of cancellation

When downgrade happens, it's possible that `http11ConnectionWaiter != null && http2ConnectionWaiter != null`. In that case `http2ConnectionWaiter.ConnectionCancellationTokenSource` will be `null`, so we never need to cancel both. It didn't seem to worth explicitly nulling out `http2ConnectionWaiter` in case of downgrade. IMO we should just rename the method as you suggest and add a comment.

stephentoub commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:
```diff
@@ -2425,5 +2481,31 @@ public QueueItem PeekNextRequestForConnectionAttempt()
+        {
+            // When a connection attempt is pending, reference the connection's CTS, so we can tear it down if the initiating request is cancelled
+            // or completes on a different connection.
+            public volatile CancellationTokenSource? ConnectionCancellationTokenSource;
```
This shouldn't need to be volatile if you just move the setting of it to null to be in the lock

stephentoub reviewed (approved):
Thanks.  It'd be good to do some stress testing on this.

antonfirsov commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:
```diff
@@ -451,27 +451,43 @@ private async Task AddHttp11ConnectionAsync(RequestQueue<HttpConnection>.QueueIt
+            catch (Exception e)
+            {
+                connectionException = e;
+            }
```
I considered it but to me it seemed less readable. Do we generally prefer shorter forms over more expressive ones?

stephentoub replied:
I personally don't think multiple catch blocks is "more expressive".  And for me it's not about length, it's about understandability: we get an exception and we need to store either that exception or one that wraps it, which we can express concisely and readably with a single catch block and ternary, making it clear exactly what the difference in behavior is based on the kind of exception, rather than two catch blocks that need to be compared to understand what the difference between them is.

stephentoub commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:
```diff
@@ -451,27 +451,43 @@ private async Task AddHttp11ConnectionAsync(RequestQueue<HttpConnection>.QueueIt
+            }
+            finally
+            {
+                waiter.ConnectionCancellationTokenSource = null;
```
I don't understand the response.  The code is currently:
```C#
waiter.ConnectionCancellationTokenSource = null;
lock (...)
{
    ...
}
```
I'm suggesting it be:
```C#
lock (...)
{
    waiter.ConnectionCancellationTokenSource = null;
    ...
}
```
Can you help me understand how your response applies?

antonfirsov replied:
I updated my response, GH didn't refresh for some reason so I missed your comment.

antonfirsov committed:
address feedback

antonfirsov committed:
compact catch block

antonfirsov commented:
/azp run runtime-libraries-coreclr outerloop

antonfirsov committed:
Merge branch 'main' into cancel-stalled-connections

antonfirsov committed:
SkipTestException doesn't seem to work on CI

MihaZupan commented on src/libraries/System.Net.Http/tests/FunctionalTests/SocketsHttpHandlerTest.Cancellation.cs:
```diff
@@ -301,7 +301,8 @@ public async Task CancelPendingRequest_DropsStalledConnectionAttempt()
             {
-                throw new SkipTestException("This test relies on ConnectCallback, and the scenario is not yet validated for HTTP/3.");
+                // This test relies on ConnectCallback, and the scenario is not yet validated for HTTP/3.
+                return;
```
`SkipTestException` requires the test to be a `ConditionalFact` instead

antonfirsov replied:
I remembered something like this but then I've seen this test which is decorated simply with `[Fact]`:

https://github.com/dotnet/runtime/blob/aafe4d465f97fa3a0ae61b18c2887cf84085d25d/src/libraries/Common/tests/System/Net/Http/HttpClientHandlerTest.cs#L154-L178

I guess it works because of a side effect of `[SkipOnPlatform]`.

antonfirsov commented:
Addressed all the feedback.

> It'd be good to do some stress testing on this.

I did a HTTP2 & HTTP11 stress run on both Windows and Linux. There were no functional failures from bugs in the code, however I had to raise the default 16K Windows dynamic port range in order to avoid failures from port exhaustion. This is not unexpected considering that in stress tests the cancellation is happening after a short 0-100 ms time interval, meaning that many connections that were previously added to the pool are now actually aborted leaving many sockets lingering in `TIME_WAIT` an preventing new connections.

I don't think quick cancellation at this scale is a typical user scenario and if it happens, the `PENDINGCONNECTIONTIMEOUTONREQUESTCOMPLETION` knob is here to deal with it.

stephentoub commented:
> however I had to raise the default 16K Windows dynamic port range in order to avoid failures from port exhaustion

How come this wasn't necessary with .NET 5, which should have had similar cancellation behavior? Or was it necessary?

antonfirsov commented:
We had encounters of `SocketException (10061)` (see #42211), but I'm not sure about the scale. Need to compare PR vs .NET 5.0 on the machine I'm seeing this.

antonfirsov committed:
log cancellation reason

antonfirsov commented:
I can confirm that in the HttpStress scenario we have much more TIME_WAIT-s than with .NET 5.0.
@MihaZupan's analysis seems to be correct: we are cancelling extra connections because of racy completions. In fact, we have much more connection cancellations because of a request being served by another connection than because of the cancellation of the originating request.

For local stress runs, raising the timeout to 50ms is sufficient to solve the issue. I'm OK with the proposal to set it to some reasonable value, like 5 sec. Another solution we can consider is to only cancel a connection when the originating request is cancelled. In my analysis this will also resolve the pool stalling problem caused by bad servers, and it's a much less intrusive change.

antonfirsov commented:
> you can either still end up stalling requests indefinitely waiting for these bad connections

If a bad connection is not being cancelled within the originating request's timeout that means that there is at least one good connection taking the originating request. Unless I'm missing some bad scenario, it seems to me that this (these) good connection(s) will eventually serve all requests in the queue.

> potentially end up with many more connections laying around than you otherwise would

This certainly is true.

That said, I'm fine setting the default "pending connection timeout" to a non-zero value instead. The disadvantage of that solution is that it will make `ConnectTimeout` somewhat virtual for pools which are bombarded with many parallel requests, since we will end up using a significantly shorter timeout for connection attempts on those in practice. If we are unhappy with this solution, we can experiment with [other mitigation ideas from our list](https://github.com/dotnet/runtime/issues/66297#issuecomment-1143761114), but so far we were not able to come up with an idea that doesn't have any disadvantages.

@stephentoub @MihaZupan thoughts/opinions?

MihaZupan commented:
What if we apply this new cancellation only when `ConnectTimeout` is not set (it's infinite)?
- It avoids clashing with a user-specified `ConnectTimeout`
- It is similar to setting a non-infinite default for `ConnectTimeout`, but it will only cancel connections, never requests. It avoids issues like that of NuGet, as requests will still fail with the same exceptions.
- We can set the default to 15 seconds, like we were going to do for `ConnectTimeout`

antonfirsov commented:
I can imagine a surprised user discovering that connections are timing out & being destroyed in the background despite `ConnectTimeout` being infinite. Regardless, this looks like an acceptable compromise to me.

antonfirsov commented:
> It should be possible to tell whether it was a default value, or the user themselves set the infinite timeout, shouldn't it?

I don't think it matters who sets the value, the point is that despite the value being infinite, we are tearing down connections after some time. I think distinguishing "default infinite" from "user infinite" would make things only more complicated.

antonfirsov committed:
default PendingConnectionTimeoutOnRequestCompletion to 15 sec

antonfirsov committed:
update tests

antonfirsov committed:
Merge branch 'main' into cancel-stalled-connections

antonfirsov committed:
wip

antonfirsov committed:
tests

antonfirsov committed:
change default timeout to 5sec & special-case infinite timeout

antonfirsov committed:
better errors and tests

antonfirsov committed:
borrow test fix from #72615

antonfirsov commented:
/azp run runtime-libraries-coreclr outerloop

antonfirsov commented:
/azp run runtime-libraries stress-http

antonfirsov committed:
forward original OCE when cancelledByOriginatingRequestCompletion

antonfirsov committed:
delete sandbox logging code

antonfirsov commented:
/azp run runtime-libraries-coreclr outerloop

antonfirsov commented:
/azp run runtime-libraries-coreclr outerloop

antonfirsov commented:
/azp run runtime-libraries stress-http

antonfirsov committed:
actually set CancelledByOriginatingRequestCompletion

antonfirsov commented:
/azp run runtime-libraries-coreclr outerloop

antonfirsov commented:
/azp run runtime-libraries stress-http

antonfirsov committed:
simplify connection exception code

antonfirsov commented:
/azp run runtime-libraries-coreclr outerloop

antonfirsov commented:
/azp run runtime-libraries stress-http

stephentoub reviewed (approved):
Assuming CI and http stress look good, LGTM.

antonfirsov committed:
fix assertion failure (hopefully)

antonfirsov committed:
Merge branch 'main' into cancel-stalled-connections

# Conflicts:
#	src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs

antonfirsov commented:
> Assuming CI and http stress look good

https://github.com/dotnet/runtime/blob/785f375e5ca9d1a9f8771d134fa92bed4f31bdd3/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs#L1147-L1151

^ this assertion failed in one of the runs, which I addressed in dbdbbc389bd7e22e8ca8d9bc802edd1ff4dc01d7. If there is no pushback against that change and the next CI run is OK, I will merge.

antonfirsov commented:
/azp run runtime-libraries-coreclr outerloop

antonfirsov commented:
/azp run runtime-libraries stress-http

antonfirsov commented:
Test failures are unrelated, System.Net.Http failure is #72818 / #72586.
